pandas/tests/tools/test_to_datetime.py

```diff
@@ -2393,85 +2298,85 @@ def test_day_not_in_month_raise_value(self, cache, arg):
-                "20010101 1234": datetime(2001, 1, 1, 12, 34),
-                "20010101 123456": datetime(2001, 1, 1, 12, 34, 56),
-            }.items()
-        ),
+        "date_str, expected, warning",
+        [
+            ("2011-01-01", datetime(2011, 1, 1), None),
+            ("2Q2005", datetime(2005, 4, 1), UserWarning),
+            ("2Q05", datetime(2005, 4, 1), UserWarning),
+            ("2005Q1", datetime(2005, 1, 1), UserWarning),
+            ("05Q1", datetime(2005, 1, 1), UserWarning),
+            ("2011Q3", datetime(2011, 7, 1), UserWarning),
+            ("11Q3", datetime(2011, 7, 1), UserWarning),
+            ("3Q2011", datetime(2011, 7, 1), UserWarning),
+            ("3Q11", datetime(2011, 7, 1), UserWarning),
+            # quarterly without space
+            ("2000Q4", datetime(2000, 10, 1), UserWarning),
+            ("00Q4", datetime(2000, 10, 1), UserWarning),
+            ("4Q2000", datetime(2000, 10, 1), UserWarning),
+            ("4Q00", datetime(2000, 10, 1), UserWarning),
+            ("2000q4", datetime(2000, 10, 1), UserWarning),
+            ("2000-Q4", datetime(2000, 10, 1), UserWarning),
+            ("00-Q4", datetime(2000, 10, 1), UserWarning),
+            ("4Q-2000", datetime(2000, 10, 1), UserWarning),
+            ("4Q-00", datetime(2000, 10, 1), UserWarning),
+            ("00q4", datetime(2000, 10, 1), UserWarning),
+            ("2005", datetime(2005, 1, 1), None),
+            ("2005-11", datetime(2005, 11, 1), None),
+            ("2005 11", datetime(2005, 11, 1), UserWarning),
+            ("11-2005", datetime(2005, 11, 1), UserWarning),
+            ("11 2005", datetime(2005, 11, 1), UserWarning),
+            ("200511", datetime(2020, 5, 11), UserWarning),
+            ("20051109", datetime(2005, 11, 9), None),
+            ("20051109 10:15", datetime(2005, 11, 9, 10, 15), None),
+            ("20051109 08H", datetime(2005, 11, 9, 8, 0), None),
+            ("2005-11-09 10:15", datetime(2005, 11, 9, 10, 15), None),
+            ("2005-11-09 08H", datetime(2005, 11, 9, 8, 0), None),
+            ("2005/11/09 10:15", datetime(2005, 11, 9, 10, 15), None),
+            ("2005/11/09 08H", datetime(2005, 11, 9, 8, 0), None),
+            ("Thu Sep 25 10:36:28 2003", datetime(2003, 9, 25, 10, 36, 28), None),
+            ("Thu Sep 25 2003", datetime(2003, 9, 25), None),
+            ("Sep 25 2003", datetime(2003, 9, 25), None),
+            ("January 1 2014", datetime(2014, 1, 1), None),
+            # GHE10537
+            ("2014-06", datetime(2014, 6, 1), None),
+            ("06-2014", datetime(2014, 6, 1), UserWarning),
+            ("2014-6", datetime(2014, 6, 1), None),
+            ("6-2014", datetime(2014, 6, 1), UserWarning),
+            ("20010101 12", datetime(2001, 1, 1, 12), None),
+            ("20010101 1234", datetime(2001, 1, 1, 12, 34), UserWarning),
+            ("20010101 123456", datetime(2001, 1, 1, 12, 34, 56), UserWarning),
```

Comment:
Not clear whey we would warn on the above two strings

Reply:
I'll respond to all your comments tomorrow, but in the meantime just wanted to say - thank you for your detailed review, much appreciated 🙏

Reply:
They're not currently detected by

https://github.com/pandas-dev/pandas/blob/b3aa585502db257ee72e651852d8a25b5d7700fa/pandas/_libs/tslibs/parsing.pyx#L927-L944

They'd be detected if they had hours/minute/seconds separators:
```python
In [1]: guess_datetime_format('20010101 123456')

In [2]: guess_datetime_format('20010101 12:34:56')
Out[2]: '%Y%m%d %H:%M:%S'
```

IMO this is fine for now, but `guess_datetime_format` could be expanded to guess these as well

Reply:
This seems to be a loss of functionality, because in pandas 1.5.2, `pd.to_datetime("20010101 123456")` works

Reply:
It'll still work, there'll just be a warning that the format couldn't be guessed, and so the elements won't be parsed with a consistent format (just like they aren't now)

But I'd be open to expanding `guess_datetime_format` to do this too

Reply:
So I pulled this PR, and tried it:
```python
>>> pd.to_datetime("20010101 123456")
<stdin>:1: UserWarning: Could not infer format - to ensure consistent parsing, specify a format.
Timestamp('2001-01-01 12:34:56')
```
It seems that if you are parsing a single string, the issue isn't  about "consistent" parsing.  So I think this is confusing, because it's hard to understand why the format couldn't be inferred, and the "consistent" message is more applicable when parsing a list of strings, not just one.

Reply:
That's a good point, thanks - would you be OK with this then:
```python
In [1]: pd.to_datetime("20010101 123456")
Out[1]: Timestamp('2001-01-01 12:34:56')

In [2]: pd.to_datetime(["20010101 123456", "201201"])
<ipython-input-2-aa7552b7764f>:1: UserWarning: Could not infer format - to ensure consistent parsing, specify a format.
  pd.to_datetime(["20010101 123456", "201201"])
Out[2]: DatetimeIndex(['2001-01-01 12:34:56', '2001-12-20 00:00:00'], dtype='datetime64[ns]', freq=None)
```
?

Reply:
Yes and no.

For the first case, where you parse a single string, the result makes sense.  We inferred a format and gave no message.

For the second case, the warning along with the result is a bit misleading.  We did infer a format for the first string and second string, but we inferred inconsistent formats.  So maybe the message should be "Inconsistent formats inferred - to ensure consistent parsing, specify a format."

You'd have to check if that error message is appropriate in all cases.  Maybe if the format can't be inferred from a single string, then the message should be different.

Finally, I tried this, and don't understand the result:
```python
>>> pd.to_datetime(["13-05-06", "01-05-06"])
DatetimeIndex(['2006-05-13', '2006-01-05'], dtype='datetime64[ns]', freq=None)
```
The first string is inferred to be `dd-mm-yy` but the second string is then parsed as `mm-dd-yy` .  Shouldn't the second string be `2006-05-01` if the format is guessed from the first string, that would be applied to the second string?

Reply:
> We did infer a format for the first string and second string, but we inferred inconsistent formats

That's not quite right, it's more that pandas couldn't infer the format, so we defer to `dateutil` to guess each element individually

Maybe that's all the message needs to say then:
```python
Could not infer format, so each element will be parsed individually by dateutil.
To ensure parsing is consistent and as-expected, please specify a format.
```
Then this one could be emitted regardless of whether there's a single element or many

Reply:
Hmmm.  Not sure what is different unless I did something wrong in the build.  I have the right version:
```python
>>> pd.to_datetime(["13-05-06", "01-05-06"])
DatetimeIndex(['2006-05-13', '2006-01-05'], dtype='datetime64[ns]', freq=None)
>>> pd.__version__
'2.0.0.dev0+929.g1ec70db018'
>>> parsing.guess_datetime_format("13-05-06")
>>> parsing.guess_datetime_format("13-05-06") is None
True
>>> pd.to_datetime("13-05-06")
Timestamp('2006-05-13 00:00:00')
```

Reply:
> Maybe that's all the message needs to say then:
> 
> ```
> Could not infer format, so each element will be parsed individually by dateutil.
> To ensure parsing is consistent and as-expected, please specify a format.
> ```
> 
> Then this one could be emitted regardless of whether there's a single element or many

That works for me.

Reply:
> Hmmm. Not sure what is different unless I did something wrong in the build. I have the right version:
> 
> ```python
> >>> pd.to_datetime(["13-05-06", "01-05-06"])
> DatetimeIndex(['2006-05-13', '2006-01-05'], dtype='datetime64[ns]', freq=None)
> >>> pd.__version__
> '2.0.0.dev0+929.g1ec70db018'
> >>> parsing.guess_datetime_format("13-05-06")
> >>> parsing.guess_datetime_format("13-05-06") is None
> True
> >>> pd.to_datetime("13-05-06")
> Timestamp('2006-05-13 00:00:00')
> ```

So it seems that the first time you try it, you get a warning message, but the second time, you don't:
```python
>>> import pandas as pd
>>> pd.to_datetime(["13-05-06", "01-05-06"])
<stdin>:1: UserWarning: Could not infer format - to ensure consistent parsing, specify a format.
DatetimeIndex(['2006-05-13', '2006-01-05'], dtype='datetime64[ns]', freq=None)
>>> pd.to_datetime(["13-05-06", "01-05-06"])
DatetimeIndex(['2006-05-13', '2006-01-05'], dtype='datetime64[ns]', freq=None)
```
But if I run it as a script, I get the message both times.

Reply:
OK yeah, that's just the way the Python repl works with warnings, it doesn't remit them - not much we can about that

I'll fixup the warning message, thanks for your careful review! 🙏

Reply:
> OK yeah, that's just the way the Python repl works with warnings, it doesn't re-emit them - not much we can about that

I think we could get around this.  If we create a warning in `errors\__init__.py` of class `DatetimeWarning` that is a subclass of `UserWarning`, and then in `errors\__init__.py`, we add in `warnings.filter("always", DatetimeWarning)` then the warning will get emitted every time it happens.

Concern here is that you are using the python interpreter, and you get the warning on one line, but not on a second line.

Also, by creating a specific class for this particular warning, it let's people test for it with their own filter.

Reply:
Not overly keen on that, we don't do it for any other warnings

Reply:
I think this warning is a bit different, because the warning occurs due to the data that is presented to the method.  So if you call `pd.to_datetime()` with 2 different sets of data, each generating the warning for different reasons, you could possibly miss that there was an error in the second data set.

Reply:
true, that's a really good point!

Reply:
actually, we don't do this for other warnings where it arguably should be shown twice:
```python
>>> import pandas as pd
>>> df = pd.DataFrame({'a': [1,2,3], 'b': [4,5,6]})
>>> df2 = df[df['a']>1]
>>> df2['b'] = 3
<stdin>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
>>> df = pd.DataFrame({'a': [1,2,3], 'b': [4,5,6]})
>>> df2 = df[df['a']>1]
>>> df2['b'] = 3
```

Also, even if we did this, given that we'd have set a warning to always show, would users have a way to turn it off?
```python
>>> with warnings.catch_warnings():
...     warnings.filterwarnings("ignore", category=FormatInferenceWarning)
...     pd.to_datetime(["13-05-06", "01-05-07"])
... 
<stdin>:3: FormatInferenceWarning: Could not infer format, so each element will be parsed individually by `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
DatetimeIndex(['2006-05-13', '2007-01-05'], dtype='datetime64[ns]', freq=None)
```

So, not sure this needs doing

Reply:
> Also, even if we did this, given that we'd have set a warning to always show, would users have a way to turn it off?

You can use `warnings.filterwarnings()`:

Running on a modification of your branch:

```python
>>> import pandas as pd
>>> import warnings
>>> from pandas.errors import DatetimeWarning
>>> warnings.simplefilter("ignore", DatetimeWarning)
>>> pd.to_datetime(["13-05-06", "01-05-07"])
DatetimeIndex(['2006-05-13', '2007-01-05'], dtype='datetime64[ns]', freq=None)
```
without those calls:
```python
>>> import pandas as pd
>>> pd.to_datetime(["13-05-06", "01-05-07"])
<stdin>:1: DatetimeWarning: Could not infer format - to ensure consistent parsing, specify a format.
DatetimeIndex(['2006-05-13', '2007-01-05'], dtype='datetime64[ns]', freq=None)
```

> actually, we don't do this for other warnings where it arguably should be shown twice:

Maybe we should consider doing that for more of the warnings.  `SettingWithCopyWarning` is another good example - you'd like to see the warning every time you do it.

I don't fully understand why this is only an issue in the interpreter, but not in a script or in iPython.  So maybe not worth worrying about.

Reply:
That's not working for me for some reason:
```python
>>> import pandas as pd
>>> from pandas.errors import FormatInferenceWarning
>>> import warnings
>>> warnings.simplefilter("ignore", FormatInferenceWarning)
>>> pd.to_datetime(["13-05-06", "01-05-07"])
<stdin>:1: FormatInferenceWarning: Could not infer format, so each element will be parsed individually by `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
DatetimeIndex(['2006-05-13', '2007-01-05'], dtype='datetime64[ns]', freq=None)
```

Here's the diff I applied:
```diff
diff --git a/pandas/core/tools/datetimes.py b/pandas/core/tools/datetimes.py
index 2b63836119..5bbcc89a06 100644
--- a/pandas/core/tools/datetimes.py
+++ b/pandas/core/tools/datetimes.py
@@ -136,13 +136,17 @@ def _guess_datetime_format_for_array(arr, dayfirst: bool | None = False) -> str
             )
             if guessed_format is not None:
                 return guessed_format
-            warnings.warn(
-                "Could not infer format, so each element will be parsed "
-                "individually by `dateutil`. To ensure parsing is "
-                "consistent and as-expected, please specify a format.",
-                UserWarning,
-                stacklevel=find_stack_level(),
-            )
+            from pandas.errors import FormatInferenceWarning
+
+            with warnings.catch_warnings():
+                warnings.filterwarnings("always", category=FormatInferenceWarning)
+                warnings.warn(
+                    "Could not infer format, so each element will be parsed "
+                    "individually by `dateutil`. To ensure parsing is "
+                    "consistent and as-expected, please specify a format.",
+                    FormatInferenceWarning,
+                    stacklevel=find_stack_level(),
+                )
     return None
 
 
diff --git a/pandas/errors/__init__.py b/pandas/errors/__init__.py
index 89ac1c1025..bb3349b027 100644
--- a/pandas/errors/__init__.py
+++ b/pandas/errors/__init__.py
@@ -522,6 +522,11 @@ class InvalidColumnName(Warning):
     ... # InvalidColumnName: Not all pandas column names were valid Stata variable...
     """
 
+class FormatInferenceWarning(Warning):
+    """
+    ...
+    """
+
 
 class CategoricalConversionWarning(Warning):
     """
```

Anyway, I've opened https://github.com/pandas-dev/pandas/issues/50209, let's continue this conversation there 😄

Reply:
I did it differently.  I added a line to `errors/__init__.py` that has `warnings.filterwarnings("always", category=DatetimeWarning)` so that the warning is always on by default and did not use `warnings.catch_warnings()`

Reply:
Nice, thanks! Anyway, if we do that, we'll do it for other warnings too, so let's not consider it a blocker if that's OK

Reply:
Yes, I'm fine to resolve that in #50209